File: 2/longest.pl
longest(File, Answer, Debug) :-
    read_input(File, M, N, A1),
    M = M,

    change(A1, N, [], A2),
    reverse(A2, A3),

    partial_sum(A3, 0, [], PS1),
    reverse(PS1, PS2),

    first(PS2, Min),
    left_min(PS2, Min, [], LM1),
    reverse(LM1, LM2),

    % We can use PS1 here, no reverse needed.
    % RM2 instead of RM1, for consistency with LM2

    first(PS1, Max),
    right_max(PS1, Max, [], RM2),

    length1(PS2, 0, 0, L1),

    % Careful! CurrLength must be equal to L1
    length2(RM2, LM2, 0, L1, L2, Debug),

    ( Debug ->
        write("A:\n"),
        print_list(A3),
        write("Partial_sum:\n"),
        print_list(PS2),
        write("Left_min:\n"),
        print_list(LM2),
        write("Right_max:\n"),
        print_list(RM2),
        write(L1),
        write("\n"),
        write(L2),
        write("\n")
    ;
        true
    ),

    Answer is max(L1,L2).


%===============================================================================



change([], _, L, L).
change([H|T], N, Acc, L) :-
    NewH is -H-N,
    change(T, N, [NewH|Acc], L).

partial_sum([], _, L, L).
partial_sum([H|T], Sum, Acc, L) :-
    NewSum is Sum + H,
    partial_sum(T, NewSum, [NewSum|Acc], L).

first([],[]).
first([H|_],H).

left_min([], _, L, L). % need to find what happens with CurrMin
left_min([H|T], CurrMin, Acc, L) :-
    NewMin is min(H, CurrMin),
    left_min(T, NewMin, [NewMin|Acc], L).

right_max([], _, L, L).
right_max([H|T], CurrMax, Acc, L) :-
    NewMax is max(H, CurrMax),
    right_max(T, NewMax, [NewMax|Acc], L).

length1([], _, Length, Length).
length1([H|T], I, CurrLength, Length) :-
    ( H >= 0 ->
        NewI is I+1,
        length1(T, NewI, NewI, Length)
    ;
        NewI is I+1,
        length1(T, NewI, CurrLength, Length)
    ).

length2([], _, _, CurrLength, Length, _) :-
    Length is CurrLength.

length2(_, [], _, CurrLength, Length, _) :-
    Length is CurrLength.

length2([H1|T1], [H2|T2], Diff, CurrLength, Length, Debug) :-
    ( H1 >= H2 ->
            ( Diff > CurrLength ->
                NewLength is Diff
            ;
                NewLength is CurrLength
            ),
            NewDiff is Diff+1,
            length2(T1, [H2|T2], NewDiff, NewLength, Length, Debug)
    ;
        NewDiff is Diff-1,
        NewLength is CurrLength,
        length2([H1|T1], T2, NewDiff, NewLength, Length, Debug)
    ),
    ( Debug ->
        write("CurrLength = "),
        write(CurrLength),
        write("\n")
    ;
        true
    ).


%===============================================================================



print_list([]) :-
    write("\n\n").
print_list([First|Rest]) :-
    write(First),
    write(" "),
    print_list(Rest).

read_input(File, M, N, L) :-
    open(File, read, Stream),
    read_line(Stream, [M, N]),
    read_line(Stream, L).

read_line(Stream, C) :-
    read_line_to_codes(Stream, Line),
    atom_codes(Atom, Line),
    atomic_list_concat(Atoms, ' ', Atom),
    maplist(atom_number, Atoms, C).

File: longest/longest.cpp
#include <fstream>
#include <iostream>
using namespace std;

/*
 * Με σχετικά απλούς αλγεβρικούς μετασχηματισμούς, έγινε αναγωγή του προβλήματος στο παρακάτω:
 * https://cs.stackexchange.com/a/129788
 * οπότε η βασική ιδέα του αλγορίθμου έχει προκύψει από το παραπάνω link.
 * Από 'κει και πέρα, η υλοποίηση του αλγορίθμου είναι προσωπική απόπειρα.
 */

int main(int argc, char **argv) {
  ifstream file;
  int M, N;

  file.open(argv[1]);
  file >> M;
  file >> N;
  int a[M];
  //cout << "M = " << M << endl;
  //cout << "N = " << N << endl;

  //int c;
  for (int i = 0; i < M; ++i) {
    file >> a[i];
    a[i] = -a[i]-N;
    //cout << "a[" << i << "] = " << a[i] << endl;
  }

  int partial_sum[M];
  int left_min[M];
  int right_max[M];

  partial_sum[0] = a[0];
  for (int i = 1; i < M; ++i) {
    partial_sum[i] = partial_sum[i-1] + a[i];
  }

  left_min[0] = partial_sum[0];
  right_max[M-1] = partial_sum[M-1];
  for (int i = 1; i < M; ++i) {
    left_min[i] = min(left_min[i-1], partial_sum[i]);
    right_max[M-i-1] = max(right_max[M-i], partial_sum[M-i-1]);
  }

  //cout << endl;
  int length = 0;
  for (int i = 0; i < M; ++i) {
    if (partial_sum[i] >= 0) length = i+1;
  }

  int i = 0, j = 0;
  while (i < M && j < M) {
    if (right_max[j] >= left_min[i]) {
    //  cout << "left_min = " << left_min[i] << " > " << "right_max = " << right_max[j] << endl;
      if (j-i > length) length = j-i;
      ++j;
    }
    else {
      ++i;
    }
  }

  cout << length << endl;
}
